File: python/src/pipelines/zebra_wings/embeddings/embeddings_model.py
import keras
import numpy as np
import src.pipelines.zebra_wings.embeddings.presaved_embeddings_initializer as pei
import src.pipelines.zebra_wings.embeddings.combine_ngrams_layer as ctg


class EmbeddingsModel:
    @staticmethod
    def setup_model(pretrained_ngram_vectors, max_words, max_ngrams, embedding_dimensions):
        model = keras.models.Sequential()
        embed = keras.layers.Embedding(
            len(pretrained_ngram_vectors),
            embedding_dimensions,
            mask_zero=True,
            input_length=max_ngrams,
            trainable=False,
            embeddings_initializer=pei.PreSavedEmbeddingsInitializer(pretrained_ngram_vectors)
        )
        model.add(keras.layers.TimeDistributed(
            embed, input_shape=(max_words, max_ngrams)))
        model.add(ctg.CombineNgramsLayer())
        return model

    def __init__(
        self,
        ngram_to_id_dictionary,
        max_words, max_ngrams,
        embedding_dimensions,
        tokenizer,
        pretrained_embedding_model=None,
        pretrained_ngram_vectors=None
    ):
        self.__ngram_to_id_dictionary = ngram_to_id_dictionary
        # self.__max_chars_per_word = max_chars_per_word
        self.__max_words = max_words
        self.__max_ngrams = max_ngrams
        self.__embedding_dimensions = embedding_dimensions
        self.__model = pretrained_embedding_model if pretrained_embedding_model != None else EmbeddingsModel.setup_model(
            pretrained_ngram_vectors, self.__max_words, self.__max_ngrams, self.__embedding_dimensions,
        )
        self.__model_input = None
        self.tokenizer = tokenizer

    def keras_model(self): return self.__model

    def model_input(self):
        if not self.__model_input:
            _input = keras.layers.Input(shape=(self.__max_words, self.__max_ngrams),)
            embedded = self.__model(_input)
            self.__model_input = keras.models.Model(inputs=_input, outputs=embedded)
        return self.__model_input


    def embed(self, sentences):
        sentences_tensor = self.sentence_to_word_ids(sentences)
        return self.model_input().predict_on_batch(sentences_tensor)

    def dictionary(self):
        return self.__ngram_to_id_dictionary

    def embed_by_word_characters(self, sentences):
        sentences_tensor = self.sentence_to_char_ids(sentences)
        return self.model_input().predict_on_batch(sentences_tensor)

    def sentence_to_char_ids(self, sentences):
        sentences_splitted_by_words = list(
            map(lambda s: self.tokenizer.split_sentence_to_words(s), sentences))
        buffer = np.zeros((len(sentences), self.__max_words, self.__max_ngrams), dtype=np.int32)
        for si, sentence in enumerate(sentences_splitted_by_words):
            for wi, word in enumerate(sentence):

                for li, letter in enumerate(list(word)):
                    if (li >= self.__max_ngrams):
                        break    
                    if letter in self.__ngram_to_id_dictionary:
                        buffer[si, wi, li] = self.__ngram_to_id_dictionary[letter]
                    else:
                        buffer[si, wi, li] = 0
        return buffer

    def sentence_to_word_ids(self, sentences):
        sentences_splitted_by_words = list(
            map(lambda s: self.tokenizer.split_sentence_to_words(s), sentences))
        buffer = np.zeros((len(sentences), self.__max_words,
                           self.__max_ngrams), dtype=np.int32)
        for si, sentence in enumerate(sentences_splitted_by_words):
            for wi, word in enumerate(sentence):
                if word in self.__ngram_to_id_dictionary:
                    buffer[si, wi, 0] = self.__ngram_to_id_dictionary[word]
                else:
                    grams = self.generate_word_ids_from_ngrams(word)
                    for gi, gram in enumerate(grams):
                        if (gi >= self.__max_ngrams):
                            print(['Word exceeding max n grams per word: ', word])
                        else:
                            buffer[si, wi, gi] = gram
        return buffer

    def generate_word_ids_from_ngrams(self, word):
        vec_ids = []
        add_to_vecs_if_not_present = (
            lambda ngram: (
                False if ngram not in self.__ngram_to_id_dictionary
                else vec_ids.append(self.__ngram_to_id_dictionary[ngram]) == None and True
            )
        )
        # first try using ngrams to reconstruct the word vector
        if len(word) > 2:
            all_ngrams_found = True
            word_ngrams = self.tokenizer.split_word_to_bigrams(word)
            for wt in word_ngrams:
                if all_ngrams_found and not add_to_vecs_if_not_present(wt):
                    all_ngrams_found = False
            if all_ngrams_found:
                return vec_ids
        vec_ids = []
        # if not by ngrams use characters to construct the word vector
        # TODO: use characters to construct ngrams, not the word
        for char in list(word):
            add_to_vecs_if_not_present(char)
        return vec_ids

File: typescript/src/pipelines/zebraWings/embeddings/EmbeddingsModel.ts
import * as tf from '@tensorflow/tfjs';
import * as types from '../../../types';
import { CombineNgramsLayer } from './CombineNgramsLayer';
import { PreSavedEmbeddingsInitializer } from './PreSavedEmbeddingsInitializer';

export class EmbeddingsModel {
    public static setupModel(
        pretrainedNGramVectors: types.PretrainedDict,
        maxWords: number,
        maxNgrams: number,
        embeddingDimensions: number
    ) {
        const model = tf.sequential();
        const embedLayer = tf.layers.embedding({
            embeddingsInitializer: new PreSavedEmbeddingsInitializer({
                embeddingDimensions,
                pretrainedNGramVectors
            }),
            inputDim: pretrainedNGramVectors.size,
            inputLength: [maxNgrams],
            maskZero: true,
            outputDim: embeddingDimensions,
            trainable: false
        });
        model.add(tf.layers.timeDistributed({ layer: embedLayer, inputShape: [maxWords, maxNgrams] }));
        model.add(new CombineNgramsLayer({}));
        return model;
    }

    public tokenizer: types.IAidaTokenizer;

    private ngramToIdDictionary: { [key: string]: number };
    private maxWords: number;
    private maxNgrams: number;
    private embeddingDimensions: number;
    private model: tf.Model;

    constructor(
        ngramToIdDictionary: { [key: string]: number },
        maxWords: number,
        maxNgrams: number,
        embeddingDimensions: number,
        tokenizer: types.IAidaTokenizer,
        pretrainedEmbeddingModel?: tf.Model,
        pretrainedNGramVectors?: types.PretrainedDict
    ) {
        this.ngramToIdDictionary = ngramToIdDictionary;
        this.maxWords = maxWords;
        this.maxNgrams = maxNgrams;
        this.embeddingDimensions = embeddingDimensions;
        this.model = pretrainedEmbeddingModel
            ? pretrainedEmbeddingModel
            : EmbeddingsModel.setupModel(pretrainedNGramVectors || new Map(), this.maxWords, this.maxNgrams, this.embeddingDimensions);
        this.tokenizer = tokenizer;
    }

    public tfModel = () => this.model;

    // Embeds by word bigrams
    public embed = (sentences: string[]) => {
        return tf.tidy(() => {
            const maxWords = this.maxWords;
            const maxNgrams = this.maxNgrams;
            const input = tf.layers.input({ shape: [maxWords, maxNgrams], dtype: 'int32' });
            const embedded = this.model.apply(input) as tf.SymbolicTensor;
            const entryModel = tf.model({ inputs: input, outputs: embedded });
            const sentencesTensor = this.sentencesToWordIds(sentences);
            const output = entryModel.predictOnBatch(sentencesTensor) as tf.Tensor<tf.Rank.R3>;
            sentencesTensor.dispose();
            return output;
        });
    };

    public dictionary = () => this.ngramToIdDictionary;

    public embedByWordCharacters = (sentences: string[]) => {
        return tf.tidy(() => {
            const maxWords = this.maxWords;
            const maxNgrams = this.maxNgrams;
            const input = tf.layers.input({ shape: [maxWords, maxNgrams], dtype: 'int32' });
            const embedded = this.model.apply(input) as tf.SymbolicTensor;
            const entryModel = tf.model({ inputs: input, outputs: embedded });
            const sentencesTensor = this.sentencesToCharIds(sentences);
            const output = entryModel.predictOnBatch(sentencesTensor) as tf.Tensor<tf.Rank.R3>;
            sentencesTensor.dispose();
            return output;
        });
    };

    private sentencesToCharIds = (sentences: string[]) => {
        return tf.tidy(() => {
            const sentencesSplittedByWords = sentences.map(s => this.tokenizer.splitSentenceToWords(s));
            const buffer = tf.buffer([sentences.length, this.maxWords, this.maxNgrams], 'int32') as tf.TensorBuffer<tf.Rank.R3>;
            sentencesSplittedByWords.forEach((s, sentenceIndex) => {
                s.forEach((w: string, wordIndex: number) => {
                    w.split('').forEach((letter, lidx) => {
                        if (lidx >= this.maxNgrams) {
                            return;
                        }
                        if (this.ngramToIdDictionary[letter] !== undefined) {
                            buffer.set(this.ngramToIdDictionary[letter], sentenceIndex, wordIndex, lidx);
                        } else {
                            buffer.set(0, sentenceIndex, wordIndex, lidx);
                        }
                    });
                });
            });
            return buffer.toTensor();
        });
    };

    private sentencesToWordIds = (sentences: string[]) => {
        return tf.tidy(() => {
            const sentencesSplittedByWords = sentences.map(s => this.tokenizer.splitSentenceToWords(s));
            const buffer = tf.buffer([sentences.length, this.maxWords, this.maxNgrams], 'int32') as tf.TensorBuffer<tf.Rank.R3>;
            sentencesSplittedByWords.forEach((s, sentenceIndex) => {
                s.forEach((w: string, wordIndex: number) => {
                    if (this.ngramToIdDictionary[w] !== undefined) {
                        // use the word dictionary
                        buffer.set(this.ngramToIdDictionary[w], sentenceIndex, wordIndex, 0);
                    } else if (w.length) {
                        this.generateWordIdsFromNGrams(w).forEach((gram, gramIndex) => {
                            if (gramIndex > this.maxNgrams) {
                                // tslint:disable-next-line:no-console
                                console.warn('Word exceeding max n grams per word: ', w);
                                return;
                            }
                            buffer.set(gram, sentenceIndex, wordIndex, gramIndex);
                        });
                    }
                });
            });
            return buffer.toTensor();
        });
    };

    private generateWordIdsFromNGrams = (word: string): number[] => {
        let vecIds: number[] = [];
        const addToVecsIfNotUndefined = (k: string) => {
            if (this.ngramToIdDictionary[k] === undefined) {
                return false;
            }
            vecIds.push(this.ngramToIdDictionary[k]);
            return true;
        };
        // first try using ngrams to reconstruct the word vector
        if (word.length > 2) {
            let allNgramsFound = true;
            const wordNgrams = this.tokenizer.splitWordToBiGrams(word);
            wordNgrams.forEach(wt => {
                if (!addToVecsIfNotUndefined(wt) && allNgramsFound) {
                    allNgramsFound = false;
                }
            });
            if (allNgramsFound) {
                return vecIds;
            }
        }
        // if not by ngrams use characters to construct the word vector
        vecIds = [];
        // TODO: use characters to construct ngrams, not the word
        word.split('').forEach(addToVecsIfNotUndefined);
        return vecIds;
    };
}
